consensus/src/experimental/buffer.rs

```diff
@@ -0,0 +1,242 @@
+// Copyright (c) The Diem Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::experimental::hashable::Hashable;
+use diem_crypto::HashValue;
+use std::collections::HashMap;
+
+pub struct LinkedItem<T: Hashable> {
+    // use option so we don't need T to be cloneable
+    elem: Option<T>,
+    // index is for find_element_by_key to have a starting position (similar to find_element)
+    index: u64,
+    next: Option<HashValue>,
+}
+
+pub type Link = Option<HashValue>;
+
+/// Buffer implementes an ordered dictionary
+/// It supports push_back, pop_front, and lookup by HashValue
+pub struct Buffer<T: Hashable> {
+    map: HashMap<HashValue, LinkedItem<T>>,
+    count: u64,
+    head: Link,
+    tail: Link,
+}
+
+/// This function compares two links if they are not None
+/// It returns true if and only if both aare not None and their keys are the same
+pub fn link_eq(link_a: &Link, link_b: &Link) -> bool {
+    link_a.is_some() && link_b.is_some() && link_a.as_ref().unwrap() == link_b.as_ref().unwrap()
+}
+
+impl<T: Hashable> Buffer<T> {
+    pub fn new() -> Self {
+        Self {
+            map: HashMap::new(),
+            count: 0,
+            head: None,
+            tail: None,
+        }
+    }
+
+    pub fn get_head(&self) -> &Link {
```

Comment:
to avoid confusing with get/set, we should use head_link/tail_link to indicate this doesn't return the element

Reply:
sounds great.